components/tikv_util/src/codec/stream_event.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+use crate::{codec::Result, Either};
+use bytes::{Buf, Bytes};
+use std::io::prelude::*;
+use std::io::Cursor;
+
+pub trait Iterator {
+    fn next(&mut self) -> Result<()>;
+
+    fn valid(&self) -> bool;
+
+    fn key(&self) -> &[u8];
+
+    fn value(&self) -> &[u8];
+}
+
+pub struct EventIterator {
+    buf: Cursor<Vec<u8>>,
+    index: usize,
+    len: usize,
+    key: Vec<u8>,
+    val: Vec<u8>,
+}
+
+impl EventIterator {
+    pub fn new(buf: Vec<u8>) -> EventIterator {
+        let len = buf.len();
+        EventIterator {
+            buf: Cursor::new(buf),
+            index: 0,
+            len,
+            key: vec![],
+            val: vec![],
+        }
+    }
+}
+
+impl Iterator for EventIterator {
+    fn next(&mut self) -> Result<()> {
+        if self.valid() {
+            let len = self.buf.get_u32_le() as usize;
+            self.index += 4;
+            self.key.resize(len, 0);
+            self.buf.read_exact(self.key.as_mut_slice())?;
+            self.index += len;
+
+            let len = self.buf.get_u32_le() as usize;
+            self.index += 4;
+            self.val.resize(len, 0);
+            self.buf.read_exact(self.val.as_mut_slice())?;
+            self.index += len;
```

Comment:
rather than copying the data into the `key` and `val` vectors, i would just keep the start..end range of `key` and `val`, ...